checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java

```diff
@@ -346,11 +352,49 @@ private void refineNeqArrayLength(
         UBQualifier S = right.minusOffset(n.getLeftOperand(), atypeFactory);
 
         UBQualifier glb = T.glb(S);
+        if (!(left.isUnknownOrBottom() || right.isUnknownOrBottom())) {
+            // If expression i has type @LTLengthOf(value = "f2", offset = "f1.length") int and
+            // expression j is less than or equal to the length of f1, then the type of i + j is
+            // @LTLengthOf("f2").
+            UBQualifier r = removeArrayLengths((LessThanLengthOf) left, (LessThanLengthOf) right);
+            glb = glb.glb(r);
+            UBQualifier l = removeArrayLengths((LessThanLengthOf) right, (LessThanLengthOf) left);
+            glb = glb.glb(l);
+        }
 
         return createTransferResult(n, in, glb);
     }
 
     /**
+     * Return the result of adding i to j,when expression i has type @LTLengthOf(value = "f2",
+     * offset = "f1.length") int and expression j is less than or equal to the length of f1, then
+     * the type of i + j is @LTLengthOf("f2").
+     *
+     * <p>Return the result of adding i to j,when expression i has type @LTLengthOf(value = "f2",
+     * offset = "f1.length - 1") int and expression j is less than the length of f1, then the type
+     * of i + j is @LTLengthOf("f2").
+     *
+     * @param i expression added to j
+     * @param j expression added to i
+     * @return the type of i + j
+     */
+    private UBQualifier removeArrayLengths(LessThanLengthOf i, LessThanLengthOf j) {
+        List<String> lessThan = new ArrayList<>();
+        List<String> lessThanOrEqaul = new ArrayList<>();
+        for (String array : i.getArrays()) {
+            if (i.isLessThanLengthOf(array)) {
+                lessThan.add(array);
+            } else if (i.hasArrayWithOffsetNeg1(array)) {
+                lessThanOrEqaul.add(array);
+            }
+        }
+        UBQualifier lessThanEqQ = j.removeArrayLengthAccess(lessThanOrEqaul);
+        UBQualifier lessThanQ = j.removeArrayLengthAccessAndNeg1(lessThan);
+
+        return lessThanEqQ.glb(lessThanQ);
```

Comment:
How this works is non-obvious when reading the code. I think a comment in this method somewhere explaining how it's accomplished would be helpful (the javadoc comment does a good job explaining what is done, but the how also deserves an explanation).